bytesize transcript gitpod (#1515)
nf-core/nf-co.re: markdown/events/2022/bytesize-gitpod.md

[4:29](https://www.youtube.com/watch?v=kBoC6QBU-M0t=269)
We're going to just go straight into the live demo. I'll explain exactly what it looks like when we go into this environment, how you can add extensions into the environment. You can add different add-ons. This is a picture from Gitpod. You can add Docker or you can add lots of different functionality within your containerized environments to test your code. I will explain how to use Git within this and also opening files like PDFs etc and live rendering for website development.

[5:02](https://www.youtube.com/watch?v=kBoC6QBU-M0t=302)

> What we're going to do now is, we're just going to go straight into how to use Gitpod. Hopefully everyone can see this. The example pipeline we're going to use is the nf-core test pipeline. I've already got the browse extension so I actually have this link already. I could also copy the link address, open up a new tab and just paste it in. But for speed I'll just click on it. So when you log in, all you need to do... and at this point it would ask you for your credentials. For me it's not because I'm already logged in. This is going straight from that repository into Gitpod. Along on the other side you can see the explorer set says, these are all the files within your repository; you have a window here to look at visualization of code or PDFs etc and then you have a terminal at the bottom. You also should have extra buttons here which I can't see. Anyway, there's all the things you would normally want to do on your own computer, you can also do here.

Reviewer: ```suggestion
What we're going to do now is, we're just going to go straight into how to use Gitpod. Hopefully, everyone can see this. The example pipeline we're going to use is the nf-core test pipeline. I've already got the browse extension so I actually have this link already. I could also copy the link address, open up a new tab and just paste it in. But for speed, I'll just click on it. So when you log in, all you need to do... and at this point, it would ask you for your credentials. For me, it's not because I'm already logged in. This is going straight from that repository into Gitpod. On the other side you can see the explorer set says, these are all the files within your repository; you have a window here to look at visualization of code or PDFs etc and then you have a terminal at the bottom. You also should have extra buttons here which I can't see. Anyway, there's all the things you would normally want to do on your own computer, you can also do here.
```

[6:21](https://www.youtube.com/watch?v=kBoC6QBU-M0t=381)
Just to show you how this is actually working, the Gitpod.yaml file is not... Did I click the right one? I'll go back. This particular repository is in the master. For some pipelines (and this is a very good example), that you want to run on Gitpod, you need to make sure it's got this Gitpod.yaml file, because this is all the programs you need. If we click on that, it's showing that we have an image, so a docker file from dockerHub. The image is being pulled, that has all the code that we need and I will go a bit more into these Gitpod yamls a bit later.

[7:05](https://www.youtube.com/watch?v=kBoC6QBU-M0t=425)

> We need to be in this development branch. I also mention at this point, if you're actually making changes to nf-core code then you'll be wanting to do this in the dev branch and not on the master branch. If we click it, we should get back to a Gitpod environment, that I was expecting. That should have all the code that we need to run this particular pipeline. Sometimes this can take a few seconds. Probably because it's a live demo, it'll be more than that, but if not we can move on to different things I can explain to you. It should pop up reasonably fast. It won't? No. Okay, there we go. Now we are actually in the the correct repository that I wanted to be in, but just move that away, just to make it clearer. You can see there's a gitpod.yml file which I was expecting to see. This is the reason why we have the code brought into this environment. In the terminal, if we type in `nextflow` it finds the tool because it's already installed within this gitpod.yml file. It told us that this container contains all the things we need to run the particular piece of code.

Reviewer: ```suggestion
We need to be in this development branch. I also mention at this point, if you're actually making changes to nf-core code then you'll be wanting to do this in the dev branch and not on the master branch. If we click it, we should get back to a Gitpod environment, that I was expecting. That should have all the code that we need to run this particular pipeline. Sometimes this can take a few seconds. Probably because it's a live demo, it'll be more than that, but if not we can move on to different things I can explain to you. It should pop up reasonably fast. It won't? No. Okay, there we go. Now we are actually in the correct repository that I wanted to be in, but just move that away, just to make it clearer. You can see there's a gitpod.yml file which I was expecting to see. This is the reason why we have the code brought into this environment. In the terminal, if we type in `nextflow` it finds the tool because it's already installed within this gitpod.yml file. It told us that this container contains all the things we need to run the particular piece of code.
```

[8:18](https://www.youtube.com/watch?v=kBoC6QBU-M0t=498)

> The other thing that's quite nice here if you want to actually test this repository, there's obviously the readme. If we open this... This is obviously not rendered... There's a preview button, so we can easily click things to preview the render to look exactly the same as it would do on the web page. It also tracks, so if you're actually needing to change something in the readme you can you can add whatever you want and it will automatically change it in the preview on the right-hand side. But I'm not going to change this at the moment. If I move this across, if we look at the the readme, it tells us how we can run this piece of code and I just want to quickly show you that it's quite easy to quickly run this particular piece of code.

Reviewer: ```suggestion
The other thing that's quite nice here if you want to actually test this repository, there's obviously the readme. If we open this... This is obviously not rendered... There's a preview button, so we can easily click things to preview the render to look exactly the same as it would do on the web page. It also tracks, so if you're actually needing to change something in the readme you can add whatever you want and it will automatically change it in the preview on the right-hand side. But I'm not going to change this at the moment. If I move this across, if we look at the readme, it tells us how we can run this piece of code and I just want to quickly show you that it's quite easy to quickly run this particular piece of code.
```

[9:05](https://www.youtube.com/watch?v=kBoC6QBU-M0t=545)

> We're going to run nextflow. Remember that this is actually in a dev branch. If we run it exactly the same as this line, we would actually pull the master branch, which is not actually what we need to run. So obviously with nextflow we can just run from the main.nf file. Then we can set a profile. Here it explains how to set the profile as as normal with nextflow code. In this particular case we're going to use docker. The reason for this is that singularity is one of the few things I know that doesn't work in Gitpods. If you need to test something with singularity you'd have to do it outside. You wouldn't be able to do it within this environment. Then we could set an outdir. Everything you need to do here is very simple to do. If you want to add a new folder, we could call it results. This is the way you can add a folder manually, by clicking one of these buttons for new files on your folders and then we can just run the code... like this.

Reviewer: ```suggestion
We're going to run nextflow. Remember that this is actually in a dev branch. If we run it exactly the same as this line, we would actually pull the master branch, which is not actually what we need to run. So obviously with nextflow we can just run from the main.nf file. Then we can set up a profile. Here it explains how to set the profile as normal with nextflow code. In this particular case, we're going to use docker. The reason for this is that singularity is one of the few things I know that doesn't work in Gitpods. If you need to test something with singularity you'd have to do it outside. You wouldn't be able to do it within this environment. Then we could set an outdir. Everything you need to do here is very simple to do. If you want to add a new folder, we could call it results. This is the way you can add a folder manually, by clicking one of these buttons for new files on your folders and then we can just run the code... like this.
```

[10:11](https://www.youtube.com/watch?v=kBoC6QBU-M0t=611)

> Now we're running nextflow and this pipeline within Gitpod. I hope you can see it's quite nice that we could quickly get to run this pipeline, see how it works and begin to see the output. The resources on these machines aren't huge, there's only a few threads, so you can't do anything that's going to be using, like, eight threads for example. They only run for maximum half an hour if you have the free account, but we'll explain later, that there's an access where you can get a limited time if you need that for your particular project.

Reviewer: ```suggestion
Now we're running Nextflow and this pipeline within Gitpod. I hope you can see it's quite nice that we could quickly get to run this pipeline, see how it works and begin to see the output. The resources on these machines aren't huge, there's only a few threads, so you can't do anything that's going to be using, like, eight threads for example. They only run for a maximum of half an hour if you have the free account, but we'll explain later, that there's an access where you can get a limited time if you need that for your particular project.
```

[10:46](https://www.youtube.com/watch?v=kBoC6QBU-M0t=646)
If we just see that the program is running and we'll see all the tasks connected are running through, we can then... I'll just wait for it to finish, shouldn't take too long... This pipeline, I should have mentioned, is actually taking FASTQ reads. It's running it through FASTQC to get quality control of those particular samples and it's using MultiQC to push them all together to make a really nice plot of all the different features and the quality scores of the FASTQ reads.

[11:20](https://www.youtube.com/watch?v=kBoC6QBU-M0t=680)

> This is a quick way of running a pipeline and you can do this for any of the nf-core pipelines that have this gitpod.yaml. It's quite fun to do. It means you can just really test out a new pipeline and see exactly how it works and if it works with your data. You can drop and drag your files into this explorer area to bring files in. There's also other ways that are normally here. There'll be a file in store, I don't know why it's not here. Might be because of the screen. Anyway, you can drop and drag files in here. I'll just take that off just see if we can find it. Yes, here we go. Sorry if I have the actual full screen. You have options here that you can save files to your local machine and you can open files and do all the kind of things you would expect to do in a normal computer.

Reviewer: ```suggestion
This is a quick way of running a pipeline and you can do this for any of the nf-core pipelines that have this gitpod.yaml. It's quite fun to do. It means you can just really test out a new pipeline and see exactly how it works and if it works with your data. You can drop and drag your files into this explorer area to bring files in. There's also other ways that are normally here. There'll be a file in store, I don't know why it's not here. Might be because of the screen. Anyway, you can drop and drag files in here. I'll just take that off just see if we can find it. Yes, here we go. Sorry if I have the actual full screen. You have options here that you can save files to your local machine and you can open files and do all the kinds of things you would expect to do on a normal computer.
```

[12:06](https://www.youtube.com/watch?v=kBoC6QBU-M0t=726)

> The other thing, now it's finished. In our results file we have our results and if we look in MultiQC, for example, we have the MultiQC report. Again, you can use these preview buttons, so that you can render these particular results into a nice file where you can see everything that's going on. Here is the results of that FASTQC result in MultiQC. I think this is quite cool to know that you can you can just quickly run these pipelines.

Reviewer: ```suggestion
The other thing, now it's finished. In our results file, we have our results and if we look in MultiQC, for example, we have the MultiQC report. Again, you can use these preview buttons, so that you can render these particular results into a nice file where you can see everything that's going on. Here is the results of that FASTQC result in MultiQC. I think this is quite cool to know that you can just quickly run these pipelines.
```

[12:36](https://www.youtube.com/watch?v=kBoC6QBU-M0t=756)

> Within this particular environment you might want to change something. So maybe you're particularly interested in changing a particular module. You can still change everything in a Gitpod environment and once you change something... I'll just add... I won't actually push this to the branch, because people won't be happy. But if I add on `cat` or whatever I've changed here. If I'm changing the code here you'll notice on the left hand side, on the source control, it's noticed that one of the files within your repository has changed. If we click this button it said that we've changed the main.nf and you can now use the git functions within Gitpod to add that particular change and you can also then make a message for that particular change. Also, down the bottom you can see that we have this development branch - on the bottom left-hand side. If we click this we can also choose to create a new branch. If you've made changes to the code this is a very quick way to create a new branch and show that we can push it back to the repository and then do a pull request.

Reviewer: ```suggestion
Within this particular environment, you might want to change something. So maybe you're particularly interested in changing a particular module. You can still change everything in a Gitpod environment and once you change something... I'll just add... I won't actually push this to the branch, because people won't be happy. But if I add on `cat` or whatever I've changed here. If I'm changing the code here you'll notice on the left-hand side, on the source control, it's noticed that one of the files within your repository has changed. If we click this button it said that we've changed the main.nf and you can now use the git functions within Gitpod to add that particular change and you can also then make a message for that particular change. Also, down the bottom, you can see that we have this development branch - on the bottom left-hand side. If we click this we can also choose to create a new branch. If you've made changes to the code this is a very quick way to create a new branch and show that we can push it back to the repository and then do a pull request.
```

[13:47](https://www.youtube.com/watch?v=kBoC6QBU-M0t=827)

> I think that's everything I really wanted to show you on this very basic example. I'm going to quickly go to another example that I think is quite helpful to explore, which is actually on the website. If I go to the website repository we can also run it from here. I'll just show a few more features that I think are really interesting to see. We can actually live render the whole website and show that the whole website is changing and we can change the underlying markdown. This is quite helpful when we've been developing the website so hopefully this will quickly load and I can show you this example.

Reviewer: ```suggestion
I think that's everything I really wanted to show you in this very basic example. I'm going to quickly go to another example that I think is quite helpful to explore, which is actually on the website. If I go to the website repository we can also run it from here. I'll just show a few more features that I think are really interesting to see. We can actually live render the whole website and show that the whole website is changing and we can change the underlying markdown. This is quite helpful when we've been developing the website so hopefully, this will quickly load and I can show you this example.
```

[14:31](https://www.youtube.com/watch?v=kBoC6QBU-M0t=871)

> I will show you in this repository. We see the gitpod.yml file, this is what I showed you before. There's many things you can put in these yml files and I think the best way to learn about this is to go to the gitpod.io website or by looking at other people's gitpod.yml files. There you can see exactly what people are doing.

Reviewer: ```suggestion
I will show you this repository. We see the gitpod.yml file, this is what I showed you before. There are many things you can put in these yml files and I think the best way to learn about this is to go to the gitpod.io website or by looking at other people's gitpod.yml files. There you can see exactly what people are doing.
```

[14:56](https://www.youtube.com/watch?v=kBoC6QBU-M0t=896)

> Before I showed you image. Image was pulling a docker container, but you also can initiate tasks. You can do lots of different commands related to docker. You can open ports to expose ports so that we can look at web browser and share a particular web browser html rendering. The other thing are the VSCode extensions. These extensions are all over here and you can make sure that each repository that you make into a Gitpod environment has all the different code linting services, adding different particular pieces of code that are very useful to have each time. You can add these extensions manually in this gitpod.yml.

Reviewer: ```suggestion
Before I showed you `image`. Image was pulling a docker container, but you also can initiate tasks. You can do lots of different commands related to docker. You can open ports to expose ports so that we can look at web browser and share a particular web browser HTML rendering. The other thing is the VSCode extensions. These extensions are all over here and you can make sure that each repository that you make into a Gitpod environment has all the different code linting services, adding different particular pieces of code that are very useful to have each time. You can add these extensions manually in this gitpod.yml.
```

[15:43](https://www.youtube.com/watch?v=kBoC6QBU-M0t=943)

> You can also set all the github settings that you would need to change. Just very quickly, because I know I'm probably running over time, here in the remote browser we can set this to being a live preview. Okay, no you can't... Probably I have done something wrong. There we go! Within this repository we're now making a live view of the actual repository that's changed and then if I wanted to go through the code into the markdown I can actually change the markdown. Like I showed you before you can actually change what's happening in this live preview of the nf-core website. I think this is super useful as well.

Reviewer: ```suggestion
You can also set all the GitHub settings that you would need to change. Just very quickly, because I know I'm probably running overtime, here in the remote browser we can set this to being a live preview. Okay, no you can't... Probably I have done something wrong. There we go! Within this repository, we're now making a live view of the actual repository that's changed and then if I wanted to go through the code into the markdown I can actually change the markdown. Like I showed you before you can actually change what's happening in this live preview of the nf-core website. I think this is super useful as well.
```

[16:30](https://www.youtube.com/watch?v=kBoC6QBU-M0t=990)

> The last thing I wanted to show you was with the github button here, you can look at all the open pull requests you can click on different ones that you're interested in trying to fix. We'll get rid of this for now. It should show you the pull request and give you all the information. If it loads... hopefully it will... Anyway, normally it should come up with all the things you can do on git basically you can do within this Gitpod environment, I think this is really handy. So there we go!

Reviewer: ```suggestion
The last thing I wanted to show you was with the GitHub button here, you can look at all the open pull requests you can click on different ones that you're interested in trying to fix. We'll get rid of this for now. It should show you the pull request and give you all the information. If it loads... hopefully it will... Anyway, normally it should come up with all the things you can do on git basically you can do within this Gitpod environment, I think this is really handy. So there we go!
```

[17:04](https://www.youtube.com/watch?v=kBoC6QBU-M0t=1024)

> You can do everything that you want to do in git, but it's within this github repository and I think this is really helpful. Finally I think I showed everything I needed to show there. I just wanted to mention here... it's only a few slides left... I wanted to mention that also these Gitpod environments are used as extensive training. Sateesh presented really well all the training materials that we have two weeks ago. All of these repositories are now running a Gitpod environment, with all the the code that you need to run them. If you're new to nextflow, this is a really cool way of getting into Gitpod and nextflow, that you can have this amazing environment in the cloud with all the tools that you need.

Reviewer: ```suggestion
You can do everything that you want to do in git, but it's within this GitHub repository and I think this is really helpful. Finally, I think I showed everything I needed to show there. I just wanted to mention it here... it's only a few slides left... I wanted to mention that also these Gitpod environments are used for extensive training. Sateesh presented really well all the training materials that we have two weeks ago. All of these repositories are now running a Gitpod environment, with all the code that you need to run them. If you're new to Nextflow, this is a really cool way of getting into Gitpod and Nextflow, that you can have this amazing environment in the cloud with all the tools that you need.
```

[17:48](https://www.youtube.com/watch?v=kBoC6QBU-M0t=1068)

> The cost. The free one is really cool, it's only 50 hours per month. You have four parallel workspaces, but you have this 30 minute activity time. As a member of the nf-core github account you can find a free professional open source account. You can get this through the following link. If you click this, there's a form you need to fill in to be able to get a professional account. You can get this through the link. If you click this, there's a professional account for free, which is really nice of Gitpod. For open source developers. Just to say there's not very many threads and I think the maximum size of each repository is about 30 to 50 gigabytes, in case you're testing and you need more than that. But normally the resources are easily enough to do all these different testing things that we want to do.

Reviewer: ```suggestion
The cost. The free one is really cool, it's only 50 hours per month. You have four parallel workspaces, but you have this 30-minute activity time. As a member of the nf-core GitHub account you can find a free professional open-source account. You can get this through the following link. If you click this, there's a form you need to fill in to be able to get a professional account. You can get this through the link. If you click this, there's a professional account for free, which is really nice of Gitpod. For open source developers. Just to say there's not very many threads and I think the maximum size of each repository is about 30 to 50 gigabytes, in case you're testing and you need more than that. But normally the resources are easily enough to do all these different testing things that we want to do.
```

[18:32](https://www.youtube.com/watch?v=kBoC6QBU-M0t=1112)

> Finally, if you need help, there's lots of people in nf-core that work with gitpod and have lots of knowledge of how to run all these things. How to make the ymls for example. There's also a section on the web page relating to Gitpod and there's really great resources and videos on gitpod.io itself. I think that's everything. Thanks for listening.

Reviewer: ```suggestion
Finally, if you need help, there's lots of people in nf-core that work with gitpod and have lots of knowledge of how to run all these things. How to make the yml's for example. There's also a section on the web page relating to Gitpod and there's really great resources and videos on gitpod.io itself. I think that's everything. Thanks for listening.
```

[18:56](https://www.youtube.com/watch?v=kBoC6QBU-M0t=1136)

> Thanks Chris, that was a really fantastic talk and really exciting to hear a bit more about Gitpod again. Are there any questions? The options are to either put your questions into the chat and I can read them out, or I will just unmute the microphones so you should be able to speak up yourself.

Reviewer: ```suggestion
Thanks Chris, that was a really fantastic talk, and really exciting to hear a bit more about Gitpod again. Are there any questions? The options are to either put your questions into the chat and I can read them out, or I will just unmute the microphones so you should be able to speak up yourself.
```
